dotnet/src/Extensions/Planning.FlowPlanner/Model/FlowStep.cs

```diff
@@ -0,0 +1,241 @@
+using System.Linq;
+using System.Reflection;
+
+/// <summary>
+/// Step within a <see cref="Flow"/> which defines the step goal, available skills, required and provided variables.
+/// </summary>
+public class FlowStep
+{
+    private List<string> _requires = new();
+
+    private List<string> _provides = new();
+
+    private List<string> _passthrough = new();
+
+    private Dictionary<string, Type?> _skillTypes = new();
+
+    private Func<IKernel, Dictionary<object, string?>, IEnumerable<object>>? _skillsFactory;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="FlowStep"/> class.
+    /// </summary>
+    /// <param name="goal">The goal of step</param>
+    /// <param name="skillsFactory">The factory to get skills</param>
+    public FlowStep(string goal, Func<IKernel, Dictionary<object, string?>, IEnumerable<object>>? skillsFactory = null)
+    {
+        this.Goal = goal;
+        this._skillsFactory = skillsFactory;
+    }
+
+    /// <summary>
+    /// Deserialization only
+    /// </summary>
+    public FlowStep() : this(string.Empty)
+    {
+    }
+
+    /// <summary>
+    /// Goal of the step
+    /// </summary>
+    public string Goal { get; set; }
+
+    /// <summary>
+    /// <see cref="CompletionType"/> of the step
+    /// </summary>
+    public CompletionType CompletionType { get; set; } = CompletionType.Once;
+
+    /// <summary>
+    /// If the CompletionType is CompletionType.ZeroOrMore, this message will be used to ask the user if they want to execute the current step or skip it.
+    /// </summary>
+    public string? StartingMessage { get; set; }
+
+    /// <summary>
+    /// If the CompletionType is CompletionType.AtleastOnce or CompletionType.ZeroOrMore, this message will be used to ask the user if they want to try the step again.
+    /// </summary>
+    public string TransitionMessage { get; set; } = "Did you want to try the previous step again?";
+
+    /// <summary>
+    /// Parameters required for executing the step
+    /// </summary>
+    public List<string> Requires
+    {
+        get => this._requires;
+        set => this._requires = value;
+    }
+
+    /// <summary>
+    /// Variables to be provided by the step
+    /// </summary>
+    public List<string> Provides
+    {
+        get => this._provides;
+        set => this._provides = value;
+    }
+
+    /// <summary>
+    /// Variables to be passed through on iterations of the step
+    /// </summary>
+    public List<string> Passthrough
+    {
+        get => this._passthrough;
+        set => this._passthrough = value;
+    }
+
+    /// <summary>
+    /// Gets or sets the skill available for the current step
+    /// </summary>
+    public List<string>? Skills
+    {
+        get => this._skillTypes.Keys.ToList();
+        set
+        {
+            Dictionary<string, Type?> skills = new();
+
+            if (value != null)
+            {
+                var types = AppDomain.CurrentDomain.GetAssemblies()
```

Comment:
Just FYI today I think one missing piece for kernel is full support of dependency injection when instantiating skills.
right now the approach here is a combination of reflection with default constructors, or skill objects registered in global collection.
actually in our production case the most complex skills are created separately and injected into global skills, instead of via reflection, due to the dependencies on external services.